test/Test/Prolens/Property.hs

```diff
@@ -66,3 +67,33 @@ eqFun :: Fun Maybe Int Int -> Fun Maybe Int Int -> PropertyT IO ()
 eqFun fun1 fun2 = do
     x <- forAll genInt
     unFun fun1 x === unFun fun2 x
+
+monoidalSpec :: Spec
+monoidalSpec = describe "Monoidal" $ do
+    describe "(->)" $ do
+        it "Identity: pappend pempty id ≡ id" $ hedgehog $ do
+            f <- forAllWith (const "f") genFunction
+            x <- forAll genInt
+            y <- forAll genInt
+            pappend pempty f (x, y) === (x, f y)
+        it "Identity: pappend id pempty ≡ id" $ hedgehog $ do
+            f <- forAllWith (const "f") genFunction
+            x <- forAll genInt
+            y <- forAll genInt
+            pappend f pempty (x, y) === (f x, y)
+        it "Associativity (right)" $ hedgehog $ do
+            f <- forAllWith (const "f") genFunction
+            g <- forAllWith (const "g") genFunction
+            h <- forAllWith (const "h") genFunction
+            x <- forAll genInt
+            y <- forAll genInt
+            z <- forAll genInt
+            pappend f (pappend g h) (x, (y, z)) === (f x, (g y, h z))
```

Comment:
This looks right.
I think we could make it even more cooler, if we would check both associativity at once as well:
```haskell
pappend f (pappend g h) (x, (y, z))  === dimap (\(a, (b, c)) -> ((a, b), c))  (\((a, b), c) -> (a, (b, c)))(pappend (pappend f g) h (x, (y, z)) )
```
What do you think?

Reply:
I think that line becomes quite long.

My counterproposal is to add two helper functions

```haskell
untupleRight :: (a, (b, c)) -> (a, b, c)
untupleRight (a, (b, c)) = (a, b, c)

untupleLeft :: ((a, b), c) -> (a, b, c)
untupleLeft ((a, b), c) = (a, b, c)
```

And the test would become

```haskell
it "Associativity" $ hedgehog $ do
    f <- forAllWith (const "f") genFunction
    g <- forAllWith (const "g") genFunction
    h <- forAllWith (const "h") genFunction
    x <- forAll genInt
    y <- forAll genInt
    z <- forAll genInt
    untupleRight (pappend f (pappend g h) (x, (y, z))) === untupleLeft (pappend (pappend f g) h ((x, y), z))
```

Reply:
Sure, the helper functions are way to go. 
I don't think that `dimap` is too long to type, so I don't see why it is a problem..

Reply:
To be honest I can't get it to compile with dimap. I'm always running into issues like 

```
Couldn't match type ‘Int’ with ‘((Int, b1), c0)’
      Expected type: (Int, (Int, Int))
                     -> (((a0, b0), Int), ((Int, b1), c0))
        Actual type: ((Int, Int), Int) -> ((Int, Int), Int)
```

Maybe it has something to do with the tuple generics?

Reply:
Okay, no worries. I would look into that separately 🙂